README.rst
This package contains a stripped down version of the SpamBayes classifier, with
the following changes:

- The classifier and tokenizer code has been kept. All other code has been
  removed.
- The tokenizer has been stripped down and simplified. In particular all code
  designed specifically for email parsing has been removed.
- The ClassifierDb class has been reduced to a simple dict subclass. The custom
  pickling code has been removed, as have all database backends.
- The remaining code has been updated and made compatible with Python 3.
- An orthogonalsparse bigram (OSB) transformation has been added.
- Unicode handling has been improved.


What's it good for?
-------------------

I use sbclassifier to protect websites against contact form spam.

With a training set of a handful each of spam and non-spam messages it is
already useful. Once the training data set gets above about 20 messages of each
type I am happy to let it filter out the most obvious spam.


Usage
------

.. code::python

    import sbclassifier

    train_spam = [
        "We wholesale Masks for both adult and kids. Prices begin at $0.19 each."
        "Quickly And Effortlessly Remove Mold From Crevices In Your Home! ",
        "Good day, are you looking a good freight forwarder service in China? ",
        "I'm betting you'd like your website to generate more traffic and leads",
    ]
    train_ham = [
        "Hi, I tried to re set my password but keeps telling me email address unknown",
        "Is it possible to buy a print of the photograph on your homepage?",
        "Hello please can you close my account as soon as possible?",
        "Just wanted to say I use your website all the time, absolutely indispensible"
    ]

    # Container for the classifier data. Persist this as a pickle or write a
    # subclass that connects to a database.
    db = sbclassifier.ClassifierDb()

    classifier = sbclassifier.Classifier(db)
    for item in train_spam:
        classifier.learn(sbclassifier.tokenize_text(item), True)
    for item in train_ham:
        classifier.learn(sbclassifier.tokenize_text(item), False)

    unknown = (
        "I've helped hundreds of companies increase their traffic "
        "and I'd love to show you what my service can do for you."
    )
    probability, evidence = classifier.spamprob(sbclassifier.tokenize_text(unknown))
    print(probability)
    print(evidence)

The above script will print out::

    0.902
    [('*H*', 0.104), ('*S*', 0.908), ('can', 0.155), ('for', 0.845), ('service', 0.845), ('traffic', 0.845), ('and', 0.908)]

sbclassifier assigns 90% probability to this unknown message being spam.
It can also produce a sequence of (word, probability) pairs that reveals the
tokens that were important in this calculation.

More information
----------------

The `spambayes source repository <https://github.com/smontanaro/spambayes/>`_
contains a wealth of information on how and why the classifier works as it
works, as does the `SpamBayes wiki <http://entrian.com/sbwiki/>`_.


Copyright
---------

Copyright (C) 2002-2013 Python Software Foundation; All Rights Reserved

The Python Software Foundation (PSF) holds copyright on all material
in this project.  You may use it under the terms of the PSF license;
see LICENSE.txt.




CHANGELOG.rst
0.1.0 (released 2021-05-16)
---------------------------

- First release


tests/test_tokenizer.py
from sbclassifier.tokenizer import add_sparse_bigrams
from sbclassifier.tokenizer import tokenize_text


def test_add_sparse_bigrams():

    assert list(add_sparse_bigrams("a b c d e".split())) == [
        "a",
        "a_b",
        "a__c",
        "b",
        "b_c",
        "b__d",
        "c",
        "c_d",
        "c__e",
        "d",
        "d_e",
    ]


def test_it_normalizes_unicode():

    # ñ in two equivalent unicode representations
    n1 = "\u00f1"
    n2 = "\u006e\u0303"

    assert list(tokenize_text(n1)) == list(tokenize_text(n2))


def test_it_restricts_charset():
    s = "aàα"
    assert list(tokenize_text(s, restrict_charset=None)) == ["aàα"]
    assert list(tokenize_text(s, restrict_charset='latin1')) == ["aà?"]
    assert list(tokenize_text(s, restrict_charset='ascii')) == ["a??"]


sbclassifier/chi2.py
import math as _math
import random


def chi2Q(x2, v, exp=_math.exp, min=min):
    """Return prob(chisq >= x2, with v degrees of freedom).

    v must be even.
    """
    assert v & 1 == 0
    # XXX If x2 is very large, exp(-m) will underflow to 0.
    m = x2 / 2.0
    sum = term = exp(-m)
    for i in range(1, v // 2):
        term *= m / i
        sum += term
    # With small x2 and large v, accumulated roundoff error, plus error in
    # the platform exp(), can cause this to spill a few ULP above 1.0.  For
    # example, chi2Q(100, 300) on my box has sum == 1.0 + 2.0**-52 at this
    # point.  Returning a value even a teensy bit over 1.0 is no good.
    return min(sum, 1.0)


def normZ(z, sqrt2pi=_math.sqrt(2.0 * _math.pi), exp=_math.exp):
    "Return value of the unit Gaussian at z."
    return exp(-z * z / 2.0) / sqrt2pi


def normP(z):
    """Return area under the unit Gaussian from -inf to z.

    This is the probability that a zscore is <= z.
    """

    # This is very accurate in a fixed-point sense.  For negative z of
    # large magnitude (<= -8.3), it returns 0.0, essentially because
    # P(-z) is, to machine precision, indistiguishable from 1.0 then.

    # sum <- area from 0 to abs(z).
    a = abs(float(z))
    if a >= 8.3:
        sum = 0.5
    else:
        sum2 = term = a * normZ(a)
        z2 = a * a
        sum = 0.0
        i = 1.0
        while sum != sum2:
            sum = sum2
            i += 2.0
            term *= z2 / i
            sum2 += term

    if z >= 0:
        result = 0.5 + sum
    else:
        result = 0.5 - sum

    return result


def normIQ(p, sqrt=_math.sqrt, ln=_math.log):
    """Return z such that the area under the unit Gaussian from z to +inf is p.

    Must have 0.0 <= p <= 1.0.
    """

    assert 0.0 <= p <= 1.0
    # This is a low-accuracy rational approximation from Abramowitz & Stegun.
    # The absolute error is bounded by 3e-3.

    flipped = False
    if p > 0.5:
        flipped = True
        p = 1.0 - p

    if p == 0.0:
        z = 8.3
    else:
        t = sqrt(-2.0 * ln(p))
        z = t - (2.30753 + 0.27061 * t) / (1.0 + 0.99229 * t + 0.04481 * t ** 2)

    if flipped:
        z = -z
    return z


def normIP(p):
    """Return z such that the area under the unit Gaussian from -inf to z is p.

    Must have 0.0 <= p <= 1.0.
    """
    z = normIQ(1.0 - p)
    # One Newton step should double the # of good digits.
    return z + (p - normP(z)) / normZ(z)


def main():
    from spambayes.histogram import Hist
    import sys

    class WrappedRandom:
        # There's no way W-H is equidistributed in 50 dimensions, so use
        # Marsaglia-wrapping to shuffle it more.

        def __init__(self, baserandom=random.random, tabsize=513):
            self.baserandom = baserandom
            self.n = tabsize
            self.tab = [baserandom() for _i in range(tabsize)]
            self.next = baserandom()

        def random(self):
            result = self.next
            i = int(result * self.n)
            self.next = self.tab[i]
            self.tab[i] = self.baserandom()
            return result

    random = WrappedRandom().random
    # from uni import uni as random
    # print(random)

    def judge(ps, ln=_math.log, ln2=_math.log(2), frexp=_math.frexp):
        H = S = 1.0
        Hexp = Sexp = 0
        for p in ps:
            S *= 1.0 - p
            H *= p
            if S < 1e-200:
                S, e = frexp(S)
                Sexp += e
            if H < 1e-200:
                H, e = frexp(H)
                Hexp += e
        S = ln(S) + Sexp * ln2
        H = ln(H) + Hexp * ln2
        n = len(ps)
        S = 1.0 - chi2Q(-2.0 * S, 2 * n)
        H = 1.0 - chi2Q(-2.0 * H, 2 * n)
        return S, H, (S - H + 1.0) / 2.0

    warp = 0
    bias = 0.99
    if len(sys.argv) > 1:
        warp = int(sys.argv[1])
    if len(sys.argv) > 2:
        bias = float(sys.argv[2])

    h = Hist(20, lo=0.0, hi=1.0)
    s = Hist(20, lo=0.0, hi=1.0)
    score = Hist(20, lo=0.0, hi=1.0)

    for _i in range(5000):
        ps = [random() for _j in range(50)]
        s1, h1, score1 = judge(ps + [bias] * warp)
        s.add(s1)
        h.add(h1)
        score.add(score1)

    print("Result for random vectors of 50 probs, +", warp, "forced to", bias)

    # Should be uniformly distributed on all-random data.
    print()
    print("H", end="")
    h.display()

    # Should be uniformly distributed on all-random data.
    print()
    print("S", end="")
    s.display()

    # Distribution doesn't really matter.
    print()
    print("(S-H+1)/2",)
    score.display()


def showscore(ps, ln=_math.log, ln2=_math.log(2), frexp=_math.frexp):
    H = S = 1.0
    Hexp = Sexp = 0
    for p in ps:
        S *= 1.0 - p
        H *= p
        if S < 1e-200:
            S, e = frexp(S)
            Sexp += e
        if H < 1e-200:
            H, e = frexp(H)
            Hexp += e
    S = ln(S) + Sexp * ln2
    H = ln(H) + Hexp * ln2

    n = len(ps)
    probS = chi2Q(-2 * S, 2 * n)
    probH = chi2Q(-2 * H, 2 * n)
    print("P(chisq >= %10g | v=%3d) = %10g" % (-2 * S, 2 * n, probS))
    print("P(chisq >= %10g | v=%3d) = %10g" % (-2 * H, 2 * n, probH))

    S = 1.0 - probS
    H = 1.0 - probH
    score = (S - H + 1.0) / 2.0
    print("spam prob", S)
    print(" ham prob", H)
    print("(S-H+1)/2", score)


sbclassifier/histogram.py
import math


class Hist:
    """Simple histograms of float values."""

    def __init__(self, nbuckets=200, lo=0.0, hi=100.0):
        """
        Pass None for lo and hi and it will automatically adjust to the min
        and max values seen.
        Note:  nbuckets can be passed for backward compatibility.  The
        display() method can be passed a different nbuckets value.
        """
        self.lo, self.hi = lo, hi
        self.nbuckets = nbuckets
        self.buckets = [0] * nbuckets
        self.data = []  # the raw data points
        self.stats_uptodate = False

    # Add a value to the collection.
    def add(self, x):
        self.data.append(x)
        self.stats_uptodate = False

    def compute_stats(self, percentiles=(5, 25, 75, 95)):
        """
        Compute, and set as instance attrs:
            n         # of data points
        The rest are set iff n>0:
            min       smallest value in collection
            max       largest value in collection
            median    midpoint
            mean
            pct       list of (percentile, score) pairs
            var       variance
            sdev      population standard deviation (sqrt(variance))
        self.data is also sorted.
        """
        if self.stats_uptodate:
            return
        self.stats_uptodate = True
        data = self.data
        n = self.n = len(data)
        if n == 0:
            return
        data.sort()
        self.min = data[0]
        self.max = data[-1]
        if n & 1:
            self.median = data[n // 2]
        else:
            self.median = (data[n // 2] + data[(n - 1) // 2]) / 2.0
        # Compute mean.
        # Add in increasing order of magnitude, to minimize roundoff error.
        if data[0] < 0.0:
            temp = [(abs(x), x) for x in data]
            temp.sort()
            data = [x[1] for x in temp]
            del temp
        sum = 0.0
        for x in data:
            sum += x
        mean = self.mean = sum / n
        # Compute variance.
        var = 0.0
        for x in data:
            d = x - mean
            var += d * d
        self.var = var / n
        self.sdev = math.sqrt(self.var)
        # Compute percentiles.
        self.pct = pct = []
        for p in percentiles:
            assert 0.0 <= p <= 100.0
            # In going from data index 0 to index n-1, we move n-1 times.
            # p% of that is (n-1)*p/100.
            i = (n - 1) * p / 1e2
            if i < 0:
                # Just return the smallest.
                score = data[0]
            else:
                whole = int(i)
                frac = i - whole
                score = data[whole]
                if whole < n - 1 and frac:
                    # Move frac of the way from this score to the next.
                    score += frac * (data[whole + 1] - score)
            pct.append((p, score))

    # Merge other into self.
    def __iadd__(self, other):
        self.data.extend(other.data)
        self.stats_uptodate = False
        return self

    def get_lo_hi(self):
        self.compute_stats()
        lo, hi = self.lo, self.hi
        if lo is None:
            lo = self.min
        if hi is None:
            hi = self.max
        return lo, hi

    def get_bucketwidth(self):
        lo, hi = self.get_lo_hi()
        span = float(hi - lo)
        return span / self.nbuckets

    # Set instance var nbuckets to the # of buckets, and buckets to a list
    # of nbuckets counts.
    def fill_buckets(self, nbuckets=None):
        if nbuckets is None:
            nbuckets = self.nbuckets
        if nbuckets <= 0:
            raise ValueError("nbuckets %g > 0 required" % nbuckets)
        self.nbuckets = nbuckets
        self.buckets = buckets = [0] * nbuckets

        # Compute bucket counts.
        lo, hi = self.get_lo_hi()
        bucketwidth = self.get_bucketwidth()
        for x in self.data:
            i = int((x - lo) / bucketwidth)
            if i >= nbuckets:
                i = nbuckets - 1
            elif i < 0:
                i = 0
            buckets[i] += 1

    # Print a histogram to stdout.
    # Also sets instance var nbuckets to the # of buckets, and
    # buckts to a list of nbuckets counts, but only if at least one
    # data point is in the collection.
    def display(self, nbuckets=None, WIDTH=61):
        if nbuckets is None:
            nbuckets = self.nbuckets
        if nbuckets <= 0:
            raise ValueError("nbuckets %g > 0 required" % nbuckets)
        self.compute_stats()
        n = self.n
        if n == 0:
            return
        print("%d items; mean %.2f; sdev %.2f" % (n, self.mean, self.sdev))
        print(
            "-> <stat> min %g; median %g; max %g"
            % (self.min, self.median, self.max)
        )
        pcts = ["%g%% %g" % x for x in self.pct]
        print("-> <stat> percentiles:", "; ".join(pcts))

        lo, hi = self.get_lo_hi()
        if lo > hi:
            return

        # hunit is how many items a * represents.  A * is printed for
        # each hunit items, plus any non-zero fraction thereof.
        self.fill_buckets(nbuckets)
        biggest = max(self.buckets)
        hunit, r = divmod(biggest, WIDTH)
        if r:
            hunit += 1
        print("* =", hunit, "items")

        # We need ndigits decimal digits to display the largest bucket count.
        ndigits = len(str(biggest))

        # Displaying the bucket boundaries is more troublesome.
        bucketwidth = self.get_bucketwidth()
        whole_digits = max(len(str(int(lo))), len(str(int(hi - bucketwidth))))
        frac_digits = 0
        while bucketwidth < 1.0:
            # Incrementing by bucketwidth may not change the last displayed
            # digit, so display one more.
            frac_digits += 1
            bucketwidth *= 10.0
        format = f"%{whole_digits + 1 + frac_digits}.{frac_digits}f %{ndigits}d"

        bucketwidth = self.get_bucketwidth()
        for i in range(nbuckets):
            n = self.buckets[i]
            print(format % (lo + i * bucketwidth, n), end="")
            print("*" * ((n + hunit - 1) // hunit))


sbclassifier/classifier.py
# An implementation of a Bayes-like spam classifier.
#
# Paul Graham's original description:
#
#     http://www.paulgraham.com/spam.html
#
# A highly fiddled version of that can be retrieved from our CVS repository,
# via tag Last-Graham.  This made many demonstrated improvements in error
# rates over Paul's original description.
#
# This code implements Gary Robinson's suggestions, the core of which are
# well explained on his webpage:
#
#    http://radio.weblogs.com/0101454/stories/2002/09/16/spamDetection.html
#
# This is theoretically cleaner, and in testing has performed at least as
# well as our highly tuned Graham scheme did, often slightly better, and
# sometimes much better.  It also has "a middle ground", which people like:
# the scores under Paul's scheme were almost always very near 0 or very near
# 1, whether or not the classification was correct.  The false positives
# and false negatives under Gary's basic scheme (use_gary_combining) generally
# score in a narrow range around the corpus's best spam_cutoff value.
# However, it doesn't appear possible to guess the best spam_cutoff value in
# advance, and it's touchy.
#
# The last version of the Gary-combining scheme can be retrieved from our
# CVS repository via tag Last-Gary.
#
# The chi-combining scheme used by default here gets closer to the theoretical
# basis of Gary's combining scheme, and does give extreme scores, but also
# has a very useful middle ground (small # of msgs spread across a large range
# of scores, and good cutoff values aren't touchy).
#
# This implementation is due to Tim Peters et alia.

import math as _math

from sbclassifier.chi2 import chi2Q


class Classifier:

    use_bigrams = False
    unknown_word_prob = 0.5
    unknown_word_strength = 0.45
    minimum_prob_strength = 0.1
    ham_cutoff = 0.2
    spam_cutoff = 0.9
    max_discriminators = 150

    def __init__(self, classifier_db):
        self.db = classifier_db
        self.probcache = {}

    # spamprob using the chi-squared implementation
    # Across vectors of length n, containing random uniformly-distributed
    # probabilities, -2*sum(ln(p_i)) follows the chi-squared distribution
    # with 2*n degrees of freedom.  This has been proven (in some
    # appropriate sense) to be the most sensitive possible test for
    # rejecting the hypothesis that a vector of probabilities is uniformly
    # distributed.  Gary Robinson's original scheme was monotonic *with*
    # this test, but skipped the details.  Turns out that getting closer
    # to the theoretical roots gives a much sharper classification, with
    # a very small (in # of msgs), but also very broad (in range of scores),
    # "middle ground", where most of the mistakes live.  In particular,
    # this scheme seems immune to all forms of "cancellation disease":  if
    # there are many strong ham *and* spam clues, this reliably scores
    # close to 0.5.  Most other schemes are extremely certain then -- and
    # often wrong.
    def spamprob(
        self,
        wordstream,
        evidence=False,
        ln=_math.log,
        frexp=_math.frexp,
        LN2=_math.log(2),
    ):
        """
        Return best-guess probability that wordstream is spam.

        wordstream is an iterable object producing words.
        The return value is a float in [0.0, 1.0].

        If optional arg evidence is True, the return value is a pair
            probability, evidence
        where evidence is a list of (word, probability) pairs.
        """
        # We compute two chi-squared statistics, one for ham and one for
        # spam.  The sum-of-the-logs business is more sensitive to probs
        # near 0 than to probs near 1, so the spam measure uses 1-p (so
        # that high-spamprob words have greatest effect), and the ham
        # measure uses p directly (so that lo-spamprob words have greatest
        # effect).
        #
        # For optimization, sum-of-logs == log-of-product, and f.p.
        # multiplication is a lot cheaper than calling ln().  It's easy
        # to underflow to 0.0, though, so we simulate unbounded dynamic
        # range via frexp.  The real product H = this H * 2**Hexp, and
        # likewise the real product S = this S * 2**Sexp.
        H = S = 1.0
        Hexp = Sexp = 0

        clues = self._getclues(wordstream)
        for prob, word, record in clues:
            S *= 1.0 - prob
            H *= prob
            if S < 1e-200:  # prevent underflow
                S, e = frexp(S)
                Sexp += e
            if H < 1e-200:  # prevent underflow
                H, e = frexp(H)
                Hexp += e

        # Compute the natural log of the product = sum of the logs:
        # ln(x * 2**i) = ln(x) + i * ln(2).
        S = ln(S) + Sexp * LN2
        H = ln(H) + Hexp * LN2

        n = len(clues)
        if n:
            S = 1.0 - chi2Q(-2.0 * S, 2 * n)
            H = 1.0 - chi2Q(-2.0 * H, 2 * n)

            # How to combine these into a single spam score?  We originally
            # used (S-H)/(S+H) scaled into [0., 1.], which equals S/(S+H).  A
            # systematic problem is that we could end up being near-certain
            # a thing was (for example) spam, even if S was small, provided
            # that H was much smaller.
            # Rob Hooft stared at these problems and invented the measure
            # we use now, the simpler S-H, scaled into [0., 1.].
            prob = (S - H + 1.0) / 2.0
        else:
            prob = 0.5

        if evidence:
            clues = [(w, p) for p, w, _r in clues]
            clues.sort(key=lambda a: a[1])
            clues.insert(0, ("*S*", S))
            clues.insert(0, ("*H*", H))
            return prob, clues
        else:
            return prob

    def learn(self, wordstream, is_spam):
        """Teach the classifier by example.

        wordstream is a word stream representing a message.  If is_spam is
        True, you're telling the classifier this message is definitely spam,
        else that it's definitely not spam.
        """
        self._add_msg(wordstream, is_spam)

    def unlearn(self, wordstream, is_spam):
        """In case of pilot error, call unlearn ASAP after screwing up.

        Pass the same arguments you passed to learn().
        """
        self._remove_msg(wordstream, is_spam)

    def clear(self):
        """
        Reset the training database
        """
        self.db.clear()
        self.probcache.clear()

    def probability(self, record):
        """Compute, store, and return prob(msg is spam | msg contains word).

        This is the Graham calculation, but stripped of biases, and
        stripped of clamping into 0.01 thru 0.99.  The Bayesian
        adjustment following keeps them in a sane range, and one
        that naturally grows the more evidence there is to back up
        a probability.
        """

        spamcount = record.spamcount
        hamcount = record.hamcount

        # Try the cache first
        try:
            return self.probcache[spamcount][hamcount]
        except KeyError:
            pass

        nham = float(self.db.nham or 1)
        nspam = float(self.db.nspam or 1)

        assert hamcount <= nham, "Token seen in more ham than ham trained."
        hamratio = hamcount / nham

        assert spamcount <= nspam, "Token seen in more spam than spam trained."
        spamratio = spamcount / nspam

        prob = spamratio / (hamratio + spamratio)

        S = self.unknown_word_strength
        StimesX = S * self.unknown_word_prob

        # Now do Robinson's Bayesian adjustment.
        #
        #         s*x + n*p(w)
        # f(w) = --------------
        #           s + n
        #
        # I find this easier to reason about like so (equivalent when
        # s != 0):
        #
        #        x - p
        #  p +  -------
        #       1 + n/s
        #
        # IOW, it moves p a fraction of the distance from p to x, and
        # less so the larger n is, or the smaller s is.

        n = hamcount + spamcount
        prob = (StimesX + n * prob) / (S + n)

        # Update the cache
        try:
            self.probcache[spamcount][hamcount] = prob
        except KeyError:
            self.probcache[spamcount] = {hamcount: prob}

        return prob

    # NOTE:  Graham's scheme had a strange asymmetry:  when a word appeared
    # n>1 times in a single message, training added n to the word's hamcount
    # or spamcount, but predicting scored words only once.  Tests showed
    # that adding only 1 in training, or scoring more than once when
    # predicting, hurt under the Graham scheme.
    # This isn't so under Robinson's scheme, though:  results improve
    # if training also counts a word only once.  The mean ham score decreases
    # significantly and consistently, ham score variance decreases likewise,
    # mean spam score decreases (but less than mean ham score, so the spread
    # increases), and spam score variance increases.
    # I (Tim) speculate that adding n times under the Graham scheme helped
    # because it acted against the various ham biases, giving frequently
    # repeated spam words (like "Viagra") a quick ramp-up in spamprob; else,
    # adding only once in training, a word like that was simply ignored until
    # it appeared in 5 distinct training spams.  Without the ham-favoring
    # biases, though, and never ignoring words, counting n times introduces
    # a subtle and unhelpful bias.
    # There does appear to be some useful info in how many times a word
    # appears in a msg, but distorting spamprob doesn't appear a correct way
    # to exploit it.
    def _add_msg(self, wordstream, is_spam):
        self.probcache = {}  # nuke the prob cache
        if is_spam:
            self.db.nspam += 1
        else:
            self.db.nham += 1

        for word in set(wordstream):
            try:
                record = self.db[word]
            except KeyError:
                record = WordInfo()

            if is_spam:
                record.spamcount += 1
            else:
                record.hamcount += 1

            self.db[word] = record

        self._post_training()

    def _remove_msg(self, wordstream, is_spam):
        self.probcache = {}  # nuke the prob cache
        if is_spam:
            if self.db.nspam <= 0:
                raise ValueError("spam count would go negative!")
            self.db.nspam -= 1
        else:
            if self.db.nham <= 0:
                raise ValueError("non-spam count would go negative!")
            self.db.nham -= 1

        wordinfoget = self.db.get
        for word in set(wordstream):
            record = wordinfoget(word)
            if record is not None:
                if is_spam:
                    if record.spamcount > 0:
                        record.spamcount -= 1
                else:
                    if record.hamcount > 0:
                        record.hamcount -= 1
                if record.hamcount == 0 == record.spamcount:
                    del self.db[word]
                else:
                    self.db[word] = record

        self._post_training()

    def _post_training(self):
        """This is called after training on a wordstream.  Subclasses might
        want to ensure that their databases are in a consistent state at
        this point.  Introduced to fix bug #797890."""
        pass

    # Return list of (prob, word, record) triples, sorted by increasing
    # prob.  "word" is a token from wordstream; "prob" is its spamprob (a
    # float in 0.0 through 1.0); and "record" is word's associated
    # WordInfo record if word is in the training database, or None if it's
    # not.  No more than max_discriminators items are returned, and have
    # the strongest (farthest from 0.5) spamprobs of all tokens in wordstream.
    # Tokens with spamprobs less than minimum_prob_strength away from 0.5
    # aren't returned.
    def _getclues(self, wordstream):
        mindist = self.minimum_prob_strength

        clues = []
        push = clues.append
        for word in set(wordstream):
            tup = self._worddistanceget(word)
            if tup[0] >= mindist:
                push(tup)
        clues.sort()

        if len(clues) > self.max_discriminators:
            del clues[: -self.max_discriminators]
        # Return (prob, word, record).
        return [t[1:] for t in clues]

    def _worddistanceget(self, word):
        try:
            record = self.db[word]
        except KeyError:
            record = None
            prob = self.unknown_word_prob
        else:
            prob = self.probability(record)
        distance = abs(prob - 0.5)
        return distance, prob, word, record


class WordInfo(object):
    # A WordInfo is created for each distinct word.  spamcount is the
    # number of trained spam msgs in which the word appears, and hamcount
    # the number of trained ham msgs.
    #
    # Invariant:  For use in a classifier database, at least one of
    # spamcount and hamcount must be non-zero.
    #
    # Important:  This is a tiny object.  Use of __slots__ is essential
    # to conserve memory.
    __slots__ = "spamcount", "hamcount"

    def __init__(self):
        self.__setstate__((0, 0))

    def __repr__(self):
        return "WordInfo" + repr((self.spamcount, self.hamcount))

    def __getstate__(self):
        return self.spamcount, self.hamcount

    def __setstate__(self, t):
        self.spamcount, self.hamcount = t


class ClassifierDb(dict):
    def __init__(self):
        super().__init__(self)
        self.nspam = 0
        self.nham = 0

    def clear(self):
        super().clear()
        self.nspam = 0
        self.nham = 0
